homeassistant/components/inkbird/__init__.py

```diff
@@ -2,25 +2,55 @@
 
 from __future__ import annotations
 
+from typing import Any
+
 from inkbird_ble import INKBIRDBluetoothDeviceData
 
+from homeassistant.components.bluetooth import async_last_service_info
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import Platform
-from homeassistant.core import HomeAssistant
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
 
-from .const import CONF_DEVICE_TYPE
+from .const import CONF_DEVICE_DATA, CONF_DEVICE_TYPE, DOMAIN
 from .coordinator import INKBIRDActiveBluetoothProcessorCoordinator
 
-PLATFORMS: list[Platform] = [Platform.SENSOR]
-
 INKBIRDConfigEntry = ConfigEntry[INKBIRDActiveBluetoothProcessorCoordinator]
 
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
 
 async def async_setup_entry(hass: HomeAssistant, entry: INKBIRDConfigEntry) -> bool:
     """Set up INKBIRD BLE device from a config entry."""
+    assert entry.unique_id is not None
     device_type: str | None = entry.data.get(CONF_DEVICE_TYPE)
-    data = INKBIRDBluetoothDeviceData(device_type)
-    coordinator = INKBIRDActiveBluetoothProcessorCoordinator(hass, entry, data)
+    device_data: dict[str, Any] | None = entry.data.get(CONF_DEVICE_DATA)
+    coordinator = INKBIRDActiveBluetoothProcessorCoordinator(hass, entry)
+
+    @callback
+    def _async_device_data_changed(new_device_data: dict[str, Any]) -> None:
+        """Handle device data changed."""
+        hass.config_entries.async_update_entry(
+            entry, data={**entry.data, CONF_DEVICE_DATA: new_device_data}
+        )
+
+    data = INKBIRDBluetoothDeviceData(
+        device_type,
+        device_data,
+        coordinator.async_set_updated_data,
+        _async_device_data_changed,
+    )
+    coordinator.async_set_data(data)
```

Comment:
Why make your own method and not use `coordinator.async_set_updated_data` ?

Reply:
I guess, since you make your own constructor, you could also pass it in there if it's just for initialization.

Reply:
Its confusing since its not the coordinator data, its the data parser class here.  Maybe I should rename it

Reply:
I moved the new code in `async_setup_entry` into the coordinator and got rid of `async_set_data`